contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,309 @@
+  ): Unit = {
+    val (fs, resolvedPath) = resolvePath(path, hadoopConf)
+
+    if (overwrite) {
+      return writeActions(fs, path, actions);
+    };
+
+    val tablePath = getTablePath(resolvedPath)
+
+    if (FileNames.isDeltaFile(path)) {
+      val version = FileNames.deltaVersion(path)
+
+      if (version > 0) {
+        val prevVersion = version - 1
+        val prevPath = FileNames.deltaFile(tablePath, prevVersion)
+        getDbEntry(tablePath, prevPath) match {
+          case Some(entry) => fixDeltaLog(fs, entry)
+          case None =>
+            if (!fs.exists(prevPath)) {
+              throw new java.nio.file.FileSystemException(
+                s"previous commit ${prevPath} doesn't exist"
+              )
+            }
+          // previous commit exists in fs but not in dynamodb
+        }
+      } else {
+        getDbEntry(tablePath, path) match {
+          case Some(entry) =>
+            if (entry.complete && !fs.exists(path)) {
+              throw new java.nio.file.FileSystemException(
+                s"Old entries for ${tablePath} still exist in the database"
+              )
+            }
+          case None => ;
+        }
+      }
+
+    }
+
+    val tempPath = createTemporaryPath(resolvedPath)
+
+    val entry =
+      LogEntry(
+        tablePath,
+        resolvedPath.getName,
+        tempPath,
+        false,
+        None
+      )
+
+    logDebug(s"write: writing to temporary file ${tempPath}")
+    writeActions(fs, entry.absoluteTempPath(), actions);
+
+    // commit to dynamodb E(N, false) and exclude commitTime
+    logDebug(s"write incomplete entry to dynamodb ${entry}")
+    putDbEntry(entry, overwrite = false)
+
+    try {
+      writeCopyTempFile(fs, entry.absoluteTempPath(), resolvedPath)
+      // Commit (with overwrite=true) to DynamoDB entry E(N, true)
+      // with commitTime attribute in epoch seconds
+      writePutCompleteDbEntry(entry)
+    } catch {
+      case e: Throwable =>
+        logInfo(s"${e.getClass.getName}: ignoring recoverable error: $e")
+    }
+  }
+
+  protected def writeCopyTempFile(fs: FileSystem, src: Path, dst: Path) = {
+    copyFile(fs, src, dst)
+  }
+
+  protected def writePutCompleteDbEntry(entry: LogEntry): Unit = {
+    putDbEntry(entry.asComplete(), overwrite = true)
+  }
+
+  /*
+   * Write to db in exclusive way.
+   * Method should throw @java.nio.file.FileAlreadyExistsException if path exists in cache.
+   */
+  protected def putDbEntry(
+      logEntry: LogEntry,
+      overwrite: Boolean = false
+  ): Unit
+
+  protected def getDbEntry(
+      absoluteTablePath: Path,
+      absoluteJsonPath: Path
+  ): Option[LogEntry]
+
+  protected def getLatestDbEntry(
+      tablePath: Path
+  ): Option[LogEntry]
+
+  override def isPartialWriteVisible(path: Path): Boolean = false
+}
+
+/**
+ * The file metadata to be stored in the external db
+ */
+case class LogEntry(
+    tablePath: Path,
+    fileName: String,
+    tempPath: String,
```

Comment:
document that this is the relative path to the `tablePath`

Reply:
What about making tempPath relative to `delta_log` directory (starting with '.tmp/' prefix, instead of `delta_log/.tmp/`)? It would be more consistent with `fileName` field - as it is also relative to `delta_log` and shorter / less redundant

Reply:
yup this sounds good.